From 955d56db9d4c453fbfbc2d0717865ff10a63d502 Mon Sep 17 00:00:00 2001
From: Nick Kralevich <nnk@google.com>
Date: Wed, 11 Jul 2012 14:30:39 -0700
Subject: [PATCH] Don't call memcpy() with size > src

FORTIFY_SOURCE detected code of the form

char foo[10];
char bar[20];
memcpy(bar, foo, 20);

this code has a read buffer overflow, reading more data
from foo than it contains.

Change-Id: I7b85a2788763a839ce66a37ba84d597167c4a078
---
 mDNSPosix/mDNSUNP.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/mDNSPosix/mDNSUNP.c b/mDNSPosix/mDNSUNP.c
index 17a37c6..d20f214 100755
--- a/mDNSPosix/mDNSUNP.c
+++ b/mDNSPosix/mDNSUNP.c
@@ -115,7 +115,7 @@ struct ifi_info *get_ifi_info_linuxv6(int family, int doaliases)
                     continue;   /* already processed this interface */
                 myflags = IFI_ALIAS;
             }
-            memcpy(lastname, ifname, IFNAMSIZ);
+            strncpy(lastname, ifname, IFNAMSIZ);
             ifi = (struct ifi_info*)calloc(1, sizeof(struct ifi_info));
             if (ifi == NULL) {
                 goto gotError;
@@ -157,13 +157,13 @@ struct ifi_info *get_ifi_info_linuxv6(int family, int doaliases)
             inet_pton(family, ipv6addr, &((struct sockaddr_in6 *)ifi->ifi_netmask)->sin6_addr);
 
             /* Add interface name */
-            memcpy(ifi->ifi_name, ifname, IFI_NAME);
+            strncpy(ifi->ifi_name, ifname, IFI_NAME);
 
             /* Add interface index */
             ifi->ifi_index = index;
 
             /* Add interface flags*/
-            memcpy(ifr.ifr_name, ifname, IFNAMSIZ);
+            strncpy(ifr.ifr_name, ifname, IFNAMSIZ);
             if (ioctl(sockfd, SIOCGIFFLAGS, &ifr) < 0) {
                 if (errno == EADDRNOTAVAIL) {
                     /*
-- 
2.19.2

